In [2]:
from datascience import *
import numpy as np
from math import *
from scipy import stats
import matplotlib.pyplot as plt
%matplotlib inline

## Hypothesis Testing

In Data8.2x, you have been using simulation to conduct hypothesis testing. Now that we have completed Data8.2x, this is a good time to take a step back and reflect on hypothesis testing. 

Every hypothesis test has roughly the same structure. The following 4 steps provide a rough outline:

1) State the null and alternate hypotheses. Generally, the alternate hypothesis is what you are trying to show. Essentially, if you want to show a result, we assume the opposite is true and try to prove ourselves wrong. 

2) Determine/calculate a test statistic. See your book for a formal definition, but generally, the test statistic is any quantity that helps us evaluate our sample with respect to our null hypothesis. 

3) Determine distribution of test statistic and compute a $p$-value. If you have taken inferential statistics before, you likely computed a $z$ or $t$ statistic and used a calculator or table to compute a $p$-value. This is based on asymptotic theory of sample means/proportions. This is not the approach taken in Data8.2x. With better computing power, we can use simulation to obtain an empirical distribution of our test statistic under the null hypothesis. 

4) Conclude. For low $p$-value (generally below 0.05), we reject the null hypothesis. For high $p$-value, we fail to reject. Low $p$-value implies that our sample would be very unusual if the null hypothesis were actually true. Therefore, that is evidence that the null hypothesis is wrong. 

### Example

Let's work through an example. Suppose that in the upcoming election, Referendum A is up for approval in Colorado. You suspect that in El Paso County, more than half of eligible voters support the referendum. You collect a random sample of 200 eligible voters in El Paso County and 115 of them express support. Is there evidence to conclude that supports your suspicion? 

#### Step 1: Hypothesis

State the null and alternative hypotheses.

H0: About half of eligible voters support the referendum

HA: More than than half of eligible voters support the referendum

#### Step 2: Test Statistic

Select a test statistic and compute that test statistic for the sample.

We will define our test statistic as the proportion of people who support the referendum. Because 115 out of 200 people sampled support the referendum, our test statistic is 115/200 or .575.

#### Step 3: $p$-value

3a) If $H_0$ were true, what should the value of $\hat{p}$ be close to? In other words, if in fact, half of eligible voters support the referendum, what value should your test statistic take? 

If the null hypothesis were true, the test statistic would be close to .5, as 100 of the 200 people sampled would support the referendum.

3b) In words (and in the context of this problem), describe what the $p$-value is. 

The p-value is the probability that we reach a test statistic as likely or less likely as the one calculated. In the context of the problem, the p-value is the probability that a given sample has a proportion of referendum A supporters greater than or equal to .575. 

3c) Find the $p$-value directly and using simulation. Hint: the binomial distribution will be of use here. 

In [6]:
direct = 1-stats.binom.cdf(114,200,.5)
direct

0.020018595806698514

In [21]:
#NOTE: I used several python boxes, so don't feel the need to put everything in this box.
samps = stats.binom.rvs(200,.5,size=10000)/200
counts = np.count_nonzero(samps>=.575)
pvalue = (counts/10000)*100
print('Based on a simulation of 10000 samples of 200, the probability of receiving a sample where greater than or equal to '\
     '57.5% of the population support the referendum is ',pvalue,'%. As such, we receive a p-value of '\
     ,pvalue/100,'.')

Based on a simulation of 10000 samples of 200, the probability of receiving a sample where greater than or equal to 57.5% of the population support the referendum is  1.94 %. As such, we receive a p-value of  0.0194 .


#### Step 4: Conclude

What is your conclusion? Be sure to state your conclusion in the context of the problem.

Given our p-value of .0194 or 1.94%, we conclude that the test statistic is significant at the 5% level. As such, we reject the null hypothesis in favor of the alternate hypothesis and conclude that more than half of eligible voters in El Paso County support referendum A.

## Confidence Intervals

Construct and interpret a 95% confidence interval on $p$, the true proportion of eligible El Paso County voters who support the referendum. There are many ways to construct such an interval (bootstrap, the binomial distribution, asymptotically). Select one and implement. 

Also, compare your interval to the results of your hypothesis test. Does your interval contain the value 0.5? Why does that matter? 

In [38]:
sim = stats.binom.rvs(200,.575,size=10000)/200
lower = .505
upper = .645
p_lower = np.count_nonzero(sim<=lower)/10000
p_upper = np.count_nonzero(upper<=sim)/10000
print(p_lower,p_upper)
print('As shown by the probabilities on each tail of the distribution '\
      'being near .025, our 95% confidence interval extends from',lower,'to',upper,'.')

0.0265 0.0242
As shown by the probabilities on each tail of the distribution being near .025, our 95% confidence interval extends from 0.505 to 0.645 .


The lower end of our 95% confidence interval sits at .505, a mere .005 above the distribution of .5 predicted by the null hypothesis. However, this .005 is significant, as it confirms that the null hypothesis' .5 is outside of our confidence interval. This affirms our decision to reject the null hypothesis.